MismatchGeneration/1_mismatch_generations/Open Library Mismatches/OL-Publication-Date-Mismatch-Generation.py

```diff
@@ -0,0 +1,249 @@
+    """
+    total_rows = len(wikidata_data)
+    for index, row in tqdm(wikidata_data.iterrows(), total=total_rows, desc="Processing"):
+        external_id = row["openLibraryID"]
+        wd_date = row["publicationDate"]
+        while True:
+            try:
+                pub_date = get_first_publication_date(external_id)
+                if pub_date is None:  # Skip if pub_date is None.
+                    print(f"Publication date not found on Open Library for item {row['item']}")
+                    break
+                try:
+                    ol_formatted_date = convert_to_datetime(pub_date)
+                except Exception as e:
+                    print(f"Error converting date for item {row['item']}: {e}")
+                    continue
+                if (wd_date != ol_formatted_date):
+                    try:
+                        date_str = datetime.strptime(wd_date, "%Y-%m-%dT%H:%M:%SZ").strftime("%Y-%m-%dT%H:%M:%SZ")
+                    except ValueError as e:
+                        print(f"Error: Time converting date for item {row['item']}: {e}")
+                        continue
+                    dataframe.append({
+                        "item_id": str(row["item"]),
+                        "statement_guid": str(""),
+                        "property_id": str("P577"),
+                        "wikidata_value": str(date_str),
+                        "meta_wikidata_value": str(""), 
+                        "external_value": str(pub_date),
+                        "external_url": str(""),
+                        "type": str("statement")
+                    })      
+                break
+            except JSONDecodeError as e:
+                print(f"Error decoding JSON for item {row['item']}: {e}")
+                break
+            except RateLimitException as e:
+                if e.status_code == 403:
+                    print("Rate limited. Waiting and retrying...")
+                    time.sleep(60) # Wait for 60 seconds before retrying.
+                else:
+                    raise e
+
+
+# In[10]:
+
+
+find_mismatches(wikidata_data_1, mismatches_1)
+
+
+# In[11]:
+
+
+find_mismatches(wikidata_data_2, mismatches_2)
+
+
+# In[12]:
+
+
+find_mismatches(wikidata_data_3, mismatches_3)
+
+
+# In[21]:
+
+
+mismatches_1_df = pd.DataFrame(mismatches_1)
+
+
+# In[23]:
+
+
+# Checking mismatch accuracy.
+mismatches_1_df.head()
+
+
+# In[24]:
+
+
+mismatches_2_df = pd.DataFrame(mismatches_2)
+mismatches_3_df = pd.DataFrame(mismatches_3)
+
+
+# In[25]:
+
+
+combined_mismatches_df = pd.concat([mismatches_1_df, mismatches_2_df, mismatches_3_df], ignore_index=True)
+
+
+# In[27]:
+
+
+combined_mismatches_df.shape
+
+
+# In[28]:
+
+
+mismatch_dataframe = pd.DataFrame(combined_mismatches_df)
+
+
+# In[29]:
+
+
+mismatch_dataframe.dtypes
+
+
+# In[30]:
+
+
+check_mf_formatting(mismatch_dataframe)
+
+
+# In[31]:
+
+
+mismatch_dataframe.to_csv('openlibrary_publication_date_mismatches.csv', index=False)
```

Comment:
Probably just a minor thing but a few items might not be mismatches.  Ex: Lines 3, 23, 37, 324, etc.  The years for both sets are correct but the OpenLibrary value only has the year while the Wikidata value has month and year.  Consider comparing only years when OL only has the year value in the JSON file.

Reply:
I'm seeing line 23 at least, and this is a good point. If open library only has the year, then we shouldn't be triggering a mismatch on something that has the year and the month :)